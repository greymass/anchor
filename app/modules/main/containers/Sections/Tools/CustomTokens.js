// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import ToolsCustomTokensComponent from '../../../../../shared/components/Tools/CustomTokens';

import * as AccountsActions from '../../../../../shared/actions/accounts';
import * as CustomTokensActions from '../../../../../shared/actions/customtokens';
import * as GlobalsActions from '../../../../../shared/actions/globals';
import * as SettingsActions from '../../../../../shared/actions/settings';

class ToolsCustomTokens extends Component<Props> {
  render = () => (
    <ToolsCustomTokensComponent
      {...this.props}
    />
  )
}

function mapStateToProps(state) {
  return {
    balances: state.balances,
    blockchains: state.blockchains,
    connection: state.connection,
    customtokens: state.customtokens,
    globals: state.globals,
    settings: state.settings
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...AccountsActions,
      ...CustomTokensActions,
      ...GlobalsActions,
      ...SettingsActions
    }, dispatch)
  };
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ToolsCustomTokens));
