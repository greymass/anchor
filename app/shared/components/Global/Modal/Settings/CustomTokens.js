// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Header, Message, Modal, Segment } from 'semantic-ui-react';

import GlobalModalSettingsCustomTokenConfirm from './CustomTokens/Confirm';
import GlobalModalSettingsCustomTokenForm from './CustomTokens/Form';

const initialState = {
  account: '',
  loading: false,
  symbol: '',
  token: {},
  settings: {
    customTokens: [],
  }
};

export class GlobalModalSettingsCustomTokens extends Component<Props> {
  state = initialState;
  componentWillReceiveProps(nextProps) {
    const { globals } = nextProps;
    if (globals) {
      const { contract } = globals;
      const { account, loading, symbol } = this.state;
      const contractName = account.toLowerCase();
      const symbolName = symbol.toUpperCase();
      if (loading && contract[contractName] && contract[contractName][symbolName]) {
        const token = contract[contractName][symbolName];
        this.setState({
          loading: false,
          token
        });
      }
    }
  }
  checkToken = () => {
    const { actions, settings } = this.props;
    const { account, symbol } = this.state;
    const { getCurrencyStats } = actions;
    const contractName = account.toLowerCase();
    const symbolName = symbol.toUpperCase();
    getCurrencyStats(contractName, symbolName);
    this.setState({
      loading: true
    });
  }
  addToken = () => {
    const { actions, onClose, settings } = this.props;
    const { account, symbol } = this.state;
    actions.addCustomToken(account, symbol);
    this.setState({ name: '', token: {} }, () => {
      onClose();
    });
  }
  onChange = (e, { name, value }) => this.setState({ [name]: value.trim() });
  onClose = () => {
    this.setState(initialState);
    if (this.props.onClose) {
      this.props.onClose();
    }
  }
  render() {
    const {
      account,
      loading,
      symbol,
      token
    } = this.state;
    const {
      open,
      settings,
      t
    } = this.props;
    const {
      customTokens
    } = settings;
    return (
      <Modal
        onClose={this.onClose}
        open={open}
        size="small"
      >
        <Modal.Header>
          {t('global_modal_settings_customtoken_header')}
        </Modal.Header>
        <Modal.Content>
          <p>{t('global_modal_settings_customtoken_description')}</p>
          {(token && token.supply)
            ? (
              <GlobalModalSettingsCustomTokenConfirm
                token={token}
                onSubmit={this.addToken}
              />
            )
            : (
              <GlobalModalSettingsCustomTokenForm
                loading={loading}
                onChange={this.onChange}
                onSubmit={this.checkToken}
                ref={(c) => { this.form = c; }}
                values={this.state}
              />
            )
          }
          {(token && token.status === 'not-found')
            ? (
              <Message error icon="warning sign">
                <Header>
                  {t('global_modal_settings_customtoken_notfound_header')}
                  <Header.Subheader>
                    {t('global_modal_settings_customtoken_notfound_subheader')}
                  </Header.Subheader>
                </Header>
              </Message>
            )
            : false
          }

        </Modal.Content>
        <Modal.Actions>
          <Segment basic clearing>
            <Button
              content={t('close')}
              onClick={this.onClose}
              primary
            />
          </Segment>
        </Modal.Actions>
      </Modal>
    );
  }
}

export default translate('global')(GlobalModalSettingsCustomTokens);
