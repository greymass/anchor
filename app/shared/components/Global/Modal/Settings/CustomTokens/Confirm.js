// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Container, Table } from 'semantic-ui-react';

class GlobalModalSettingsCustomTokensConfirm extends Component<Props> {
  render() {
    const {
      onSubmit,
      t,
      token
    } = this.props;
    return (
      <React.Fragment>
        <Table definition>
          <Table.Row>
            <Table.Cell width={4}>
              {t('global_modal_settings_customtoken_confirm_issuer')}
            </Table.Cell>
            <Table.Cell>
              {token.issuer}
            </Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>
              {t('global_modal_settings_customtoken_confirm_supply')}
            </Table.Cell>
            <Table.Cell>
              {token.supply}
            </Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>
              {t('global_modal_settings_customtoken_confirm_maxsupply')}
            </Table.Cell>
            <Table.Cell>
              {token.max_supply}
            </Table.Cell>
          </Table.Row>
        </Table>
        <Container textAlign="center">
          <Button
            color="green"
            content={t('global_modal_settings_customtoken_confirm')}
            onClick={onSubmit}
          />
        </Container>
      </React.Fragment>
    );
  }
}

export default translate('global')(GlobalModalSettingsCustomTokensConfirm);
