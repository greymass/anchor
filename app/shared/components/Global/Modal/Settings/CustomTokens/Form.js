// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Container, Form, Input } from 'semantic-ui-react';

class GlobalModalSettingsCustomTokensForm extends Component<Props> {
  render() {
    const {
      loading,
      onChange,
      onSubmit,
      ref,
      t,
      values
    } = this.props;
    const {
      account,
      symbol,
    } = values;
    return (
      <Form
        loading={loading}
        onSubmit={onSubmit}
        ref={ref}
      >
        <Form.Field
          autoFocus
          control={Input}
          fluid
          label={t('global_modal_settings_customtoken_form_account_label')}
          name="account"
          onChange={onChange}
          placeholder={t('global_modal_settings_customtoken_form_account_placeholder')}
          type="text"
          value={account}
        />
        <Form.Field
          control={Input}
          fluid
          label={t('global_modal_settings_customtoken_form_symbol_label')}
          name="symbol"
          onChange={onChange}
          placeholder={t('global_modal_settings_customtoken_form_symbol_placeholder')}
          type="text"
          value={symbol}
        />
        <Container textAlign="center">
          <Button
            color="green"
            content={t('global_modal_settings_customtoken_add')}
          />
        </Container>
      </Form>
    );
  }
}

export default translate('global')(GlobalModalSettingsCustomTokensForm);
