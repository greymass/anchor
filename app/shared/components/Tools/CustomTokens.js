// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { find, findIndex } from 'lodash';

import { Button, Checkbox, Container, Header, Popup, Segment, Table } from 'semantic-ui-react';

import GlobalModalSettingsCustomToken from '../Global/Modal/Settings/CustomTokens';

class ToolsCustomTokens extends Component<Props> {
  state = {
    addingToken: false
  };
  componentDidMount() {
    this.sync();
  }
  showCustomToken = () => this.setState({ addingToken: true });
  hideCustomToken = () => this.setState({ addingToken: false });
  sync = () => {
    const { actions } = this.props;
    actions.getCustomTokens();
  };
  scan = () => {
    const {
      actions,
      connection,
      customtokens,
      settings
    } = this.props;
    const {
      chainId
    } = connection;
    let {
      tokens
    } = customtokens;

    if (tokens && tokens[chainId]) {
      tokens = tokens[chainId];
    } else {
      tokens = [];
    }

    tokens = tokens.map((token) => `${chainId}:${token.contract.toLowerCase()}:${token.symbol.toUpperCase()}`);
    actions.getCurrencyBalance(settings.account, tokens);
  };
  toggleCustomToken = (e, { checked, name }) => {
    const { actions } = this.props;
    const [chainId, contract, symbol] = name.split(':');
    if (checked) {
      actions.addCustomToken(contract, symbol);
    } else {
      actions.removeCustomToken(contract, symbol);
    }
  }
  render() {
    const {
      actions,
      balances,
      blockchains,
      connection,
      customtokens,
      globals,
      settings,
      t
    } = this.props;
    const {
      customTokens
    } = settings;
    const {
      addingToken
    } = this.state;
    let {
      tokens
    } = customtokens;
    if (tokens && tokens[connection.chainId]) {
      tokens = tokens[connection.chainId];
    } else {
      tokens = [];
    }
    if (customTokens && customTokens.length) {
      customTokens.forEach((token) => {
        const [chainId, contract, symbol] = token.split(':');
        if (findIndex(tokens, { symbol, contract }) === -1) {
          tokens.unshift({
            contract,
            custom: true,
            symbol
          });
        }
      });
    }
    let filterTokens = customTokens.filter((token) => (connection.chainId !== token.split(':')[0]));
    filterTokens = filterTokens.map((token) => (token.split(':')[2]));
    const blockchain = find(blockchains, { chainId: connection.chainId });
    return (
      <Segment color="violet" piled style={{ margin: 0 }}>
        <Header>
          {t('tools_customtokens_header')}
          <Header.Subheader>
            {t('tools_customtokens_subheader')}
          </Header.Subheader>
        </Header>
        <Container fluid>
          {(blockchain.supportedContracts && blockchain.supportedContracts.includes('customtokens'))
            ? (
              <Button
                color="green"
                content={t('tools_customtokens_scan')}
                icon="search"
                onClick={this.scan}
                size="small"
              />
            )
            : (
              <Popup
                content={t('feature_not_supported_blockchain')}
                inverted
                trigger={(
                  <Button
                    color="grey"
                    content={t('tools_customtokens_scan')}
                    icon="search"
                    size="small"
                  />
                )}
              />
            )
          }
          <Button
            color="blue"
            content={t('wallet:wallet_status_add_custom_token_action')}
            floated="right"
            icon="circle plus"
            onClick={this.showCustomToken}
            size="small"
          />
        </Container>
        <GlobalModalSettingsCustomToken
          actions={actions}
          globals={globals}
          onClose={this.hideCustomToken}
          open={addingToken}
          settings={settings}
        />
        <Table columns={5} definition striped unstackable>
          <Table.Header>
            <Table.Row>
              <Table.HeaderCell>{t('tools_customtokens_contract')}</Table.HeaderCell>
              <Table.HeaderCell>{t('tools_customtokens_symbol')}</Table.HeaderCell>
              <Table.HeaderCell>{t('tools_customtokens_balance')}</Table.HeaderCell>
              <Table.HeaderCell>{t('tools_customtokens_controls')}</Table.HeaderCell>
            </Table.Row>
          </Table.Header>
          <Table.Body>
            {([].concat(tokens)
                .filter((token) => (token.symbol !== 'EOS'))
                .filter((token) => (filterTokens.indexOf(token.symbol) === -1))
                .map((token) => {
                  const name = `${connection.chainId}:${token.contract}:${token.symbol}`;
                  const isSelected =
                    !!(settings.customTokens && settings.customTokens.indexOf(name) !== -1);
                  let balance = false;
                  if (balances && settings.account && balances[settings.account]) {
                    balance = balances[settings.account][token.symbol];
                    if (balance === undefined) {
                      balance = 0;
                    }
                  }
                  return (
                    <Table.Row key={`${connection.chainId}-${token.contract}-${token.symbol}`}>
                      <Table.Cell>
                        <Header size="small">
                          {token.contract}
                          {(token.custom) ? (
                            <Header.Subheader>
                              {t('tools_customtokens_track_custom')}
                            </Header.Subheader>
                          ) : false }
                        </Header>
                      </Table.Cell>
                      <Table.Cell>
                        {token.symbol}
                      </Table.Cell>
                      <Table.Cell>
                        <strong>
                          {balance}
                        </strong>
                      </Table.Cell>
                      <Table.Cell>
                        <Checkbox
                          label={t('tools_customtokens_track_token')}
                          defaultChecked={isSelected}
                          name={name}
                          onChange={this.toggleCustomToken}
                        />
                      </Table.Cell>
                    </Table.Row>
                  );
                })
            )}
          </Table.Body>
        </Table>
      </Segment>
    );
  }
}

export default translate('tools')(ToolsCustomTokens);
