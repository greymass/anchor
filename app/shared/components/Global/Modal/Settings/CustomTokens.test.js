import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalModalSettingsCustomTokens } from './CustomTokens';

const mockProps = {
  account: 'teamgreymass',
  loading: false,
  settings: { customTokens: [] },
  symbol: 'EOS',
  token: { EOS: { issuer: 'teamgreymass' } },
  t: text => text,
};

describe('GlobalModalSettingsCustomTokens', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalModalSettingsCustomTokens {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
