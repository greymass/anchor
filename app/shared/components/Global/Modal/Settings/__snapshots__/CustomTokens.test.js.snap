// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalModalSettingsCustomTokens renders correctly 1`] = `
<Modal
  centered={true}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  size="small"
>
  <ModalHeader>
    global_modal_settings_customtoken_header
  </ModalHeader>
  <ModalContent>
    <p>
      global_modal_settings_customtoken_description
    </p>
    <Translate(GlobalModalSettingsCustomTokensForm)
      loading={false}
      onChange={[Function]}
      onSubmit={[Function]}
      values={
        Object {
          "account": "",
          "loading": false,
          "settings": Object {
            "customTokens": Array [],
          },
          "symbol": "",
          "token": Object {},
        }
      }
    />
  </ModalContent>
  <ModalActions>
    <Segment
      basic={true}
      clearing={true}
    >
      <Button
        as="button"
        content="close"
        onClick={[Function]}
        primary={true}
      />
    </Segment>
  </ModalActions>
</Modal>
`;
